### PR TITLE
Introduce ASRecursiveUnfairLock and tests

### PR DESCRIPTION
**Pull request by Adlai-Holler**
- Adds a new C lock struct that adds recursiveness to unfair_lock but isn't much slower.
- Adds an experiment `exp_unfair_lock` that makes all `ASDN::Mutex` instances use either `os_unfair_lock` or `ASRecursiveUnfairLock` under the hood, if we're on iOS/tvOS >= 10.
- The name `Mutex` isn't totally appropriate now but it's fine. If this works well we can move it.

Here's the perf on my 2016 MBP in release mode. Take synthetic tests with a grain of salt. "Uncontested" (less contested) = 2 threads, contested = 16 threads. See `ASRecursiveUnfairLockTests.m`.

Takeaways:
  - Close to plain unfair lock performance in uncontested case.
  - 25x faster in uncontested case, 50x faster in contested case.

Data:
```
testNoLockUncontested: average: 1ms, relative standard deviation: 15.788%
testPlainUnfairLockUncontested: average: 4ms, relative standard deviation: 26.805%
testRecursiveUnfairLockUncontested: average: 5ms, relative standard deviation: 10.843%
testRecursiveMutexUncontested: average: 129ms, relative standard deviation: 3.632%

testNoLockContested: average: 4ms, relative standard deviation: 7.002%
testPlainUnfairLockContested: average: 18ms, relative standard deviation: 3.598%
testRecursiveUnfairLockContested: average: 25ms, relative standard deviation: 12.423%
testRecursiveMutexContested: average: 1248ms, relative standard deviation: 3.859%
```

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ASRecursiveUnfairLock.h:</p>
<pre><code>//
//  ASRecursiveUnfairLock.h
//  Texture
//
//  Copyright (c) 2018-present, Pinterest, Inc.  All rights reserved.
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ASRecursiveUnfairLock.m:</p>
<pre><code>//
//  ASRecursiveUnfairLock.m
//  Texture
//
//  Copyright (c) 2018-present, Pinterest, Inc.  All rights reserved.
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ASRecursiveUnfairLockTests.m:</p>
<pre><code>//
//  ASRecursiveUnfairLockTests.m
//  Texture
//
//  Copyright (c) 2018-present, Pinterest, Inc.  All rights reserved.
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/13ae4abb3c96950f8128ca5e778f3524/raw/0c85b7afa4e1c15c676d2b6b89a0175920c4387c/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/0ee08d322823c03028e1a510c9d18c65/raw/6e9043f92ec6b60f6a79e2a6aa02bc6ae811ee58/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Details/ASThread.h [hunk 1]
```diff
@@ -229,25 +234,26 @@ namespace ASDN {
...
   {
     /// Constructs a non-recursive mutex (the default).
     Mutex () : Mutex (false) {}
 
     ~Mutex () {
-      ASDISPLAYNODE_THREAD_ASSERT_ON_ERROR(pthread_mutex_destroy (&_m));
+      if (gMutex_unfair) {
+        // nop
+      } else {
+        AS_POSIX_ASSERT_NOERR(pthread_mutex_destroy (&_m));
```
Comment (Adlai-Holler) [inline Source/Details/ASThread.h]: I renamed this assert macro just because it was so dang long.

File: Source/Details/ASThread.h [hunk 2]
```diff
@@ -374,100 +422,6 @@ namespace ASDN {
...
-    }
-
-    ~ReadWriteLockWriteLocker() {
-      _lock.unlock();
-    }
-
-    // non-copyable.
-    ReadWriteLockWriteLocker(const ReadWriteLockWriteLocker&) = delete;
-    ReadWriteLockWriteLocker &operator=(const ReadWriteLockWriteLocker&) = delete;
-  };
```
Comment (Adlai-Holler) [inline Source/Details/ASThread.h]: Removed unused C++ classes from this header.

**Comment by appleguy** [comment]
This is a seminal contribution. Texture (and AsyncDisplayKit) developers have always dreamed of a faster locking primitive than the unnecessarily-fair pthread mutex, but it has felt just out of reach with so much else to do. Thank you @Adlai-Holler for jumping on this opportunity!

Some background reading for folks unfamiliar with os_unfair_lock:
https://blog.mozilla.org/nfroyd/2017/03/29/on-mutex-performance-part-1/ (discusses that pthread mutex happens to be fair on OS X / iOS platforms)
https://webkit.org/blog/6161/locking-in-webkit/ (the path that Safari / WebKit developers took to address this problem)
https://twitter.com/catfish_man/status/743252500728315905?lang=en (excitement about the new-in-iOS 10 API)
https://gist.github.com/steipete/36350a8a60693d440954b95ea6cbbafc (some hard-to-read, but more broad benchmarking comparisons with other iOS locking primitives)

File: Source/ASConfigurationInternal.h [hunk 3]
```diff
@@ -10,6 +10,8 @@
 //      http://www.apache.org/licenses/LICENSE-2.0
 //
 
+/// Note this has to be public because it's imported by public header ASThread.h =/
```
Comment (jparise) [inline Source/ASConfigurationInternal.h]: But only temporarily for the lifetime of the `exp_unfair_lock` experiment?

File: Source/Details/ASRecursiveUnfairLock.m [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ */
+#define rul_set_thread(l, t) atomic_store_explicit(&l->_thread, t, memory_order_release)
+#define rul_get_thread(l) atomic_load_explicit(&l->_thread, memory_order_acquire)
+
+void ASRecursiveUnfairLockLock(ASRecursiveUnfairLock *l)
+{
+  // Just a cache for pthread_self so that we never call it twice.
+  pthread_t s = NULL;
+  
+  // Try to lock without blocking. If we fail, check what thread owns it.
```
Comment (jparise) [inline Source/Details/ASRecursiveUnfairLock.m]: grammar nit: `s/what/which/`

File: Source/Details/ASThread.h [hunk 5]
```diff
@@ -300,23 +322,50 @@ namespace ASDN {
...
       } else {
-        pthread_mutexattr_t attr;
-        ASDISPLAYNODE_THREAD_ASSERT_ON_ERROR(pthread_mutexattr_init (&attr));
-        ASDISPLAYNODE_THREAD_ASSERT_ON_ERROR(pthread_mutexattr_settype (&attr, PTHREAD_MUTEX_RECURSIVE));
-        ASDISPLAYNODE_THREAD_ASSERT_ON_ERROR(pthread_mutex_init (&_m, &attr));
-        ASDISPLAYNODE_THREAD_ASSERT_ON_ERROR(pthread_mutexattr_destroy (&attr));
+          if (!recursive) {
+            AS_POSIX_ASSERT_NOERR(pthread_mutex_init (&_m, NULL));
+          } else {
+          // Fall back to recursive mutex.
```
Comment (jparise) [inline Source/Details/ASThread.h]: Is the indentation here off one level?

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Adlai-Holler** [comment]
Thanks Jon for reviewing. Merging!

**Comment by garrettmoon** [comment]
We should be on the lookout for deadlocks due to priority inversion when this lands. Of course if all clients and Texture itself are built correctly, no priority inversion cases will come up 🤞.

**Comment by Adlai-Holler** [comment]
We good, unfair locks protect against priority inversion! All that they lack is a guarantee that you will get the lock first if you tried to lock first:

From https://opensource.apple.com/source/libplatform/libplatform-125/include/os/lock_private.h.auto.html

```
 * @abstract
 * Replacement for the deprecated OSSpinLock. Does not spin on contention but
 * waits in the kernel to be woken up by an unlock. The opaque lock value
 * contains thread ownership information that the system may use to attempt to
 * resolve priority inversions.
```

**Comment by garrettmoon** [comment]
Amazing :)

**Pull request merged**